### PR TITLE
Add mc rb command to remove empty bucket

### LINKED ISSUE #2635: `mc rm` with --older-than flag removes buckets, too
## Expected behavior

There should be a way to delete all *objects* in a bucket using the `--older-than` flag, but not the *bucket* itself.
I would expect that this happens when running `mc rm -r --older-than 30 --force minio/bucket/`.

## Actual behavior

The bucket was deleted as well. (Just for the record, it *was* older than 30 days.)

## Steps to reproduce the behavior

1. Have a bucket older than `x` days.
2. Run `mc rm -r --older-than x --force minio/bucket/`.

Alternatively:

```
mc mb minio/bucket/
mc rm -r --older-than 0 --force minio/bucket/
```

## mc version

```
$ mc version
Version: 2018-12-27T00:37:49Z
Release-tag: RELEASE.2018-12-27T00-37-49Z
Commit-id: 4e7eeb0725abcd36b54a37636ff89f29c3238fd8
```

## System information

Fedora 29

### PR DESCRIPTION
**Pull request by poornas**
Fixes #2635 - Change mc rm --recursive behavior to
delete objects in bucket but not the
bucket itself.

### CODE REVIEW

**Comment by harshavardhana** [comment]
We should mention this as a breaking change...

File: cmd/mirror-main.go [hunk 1]
```diff
@@ -232,8 +232,8 @@ func (mj *mirrorJob) doRemove(sURLs URLs) URLs {
 	contentCh := make(chan *clientContent, 1)
 	contentCh <- &clientContent{URL: *newClientURL(sURLs.TargetContent.URL.Path)}
 	close(contentCh)
-
-	errorCh := clnt.Remove(false, contentCh)
+	isRemoveBucket := true
```
Comment (Praveenrajmani) [inline cmd/mirror-main.go]: I think we should not remove the bucket in the target if the source has zero contents. This should be `false`.
 
image

Fine if it is left intentionally to work like this.

**Comment by harshavardhana** [comment]
The intention here is bucket is never removed from `mc mirror` or `mc cp` operation. Even with `mc rm` we have decided to not remove buckets anymore. 

To remove buckets explicitly only command is `mc rb`

File: cmd/mirror-main.go (hunk 1 above)
Comment (poornas) [reply cmd/mirror-main.go]: @Praveenrajmani , changed it so mirror will not remove buckets

File: docs/minio-client-complete-guide.md [hunk 2]
```diff
@@ -314,6 +315,30 @@ mc mb s3/mybucket --region=us-west-1
 Bucket created successfully ‘s3/mybucket’.
 ```
 
+<a name="rb"></a>
+### Command `rb` - Remove a Bucket
+`rb` command removes a bucket and all its contents on an object storage. On a filesystem, it behaves like `rmdir` command.
+
+Note that when a bucket is removed all policies associated with the bucket will also be removed. If you would like to just
```
Comment (Praveenrajmani) [inline docs/minio-client-complete-guide.md]: One comment - missed a `,` here. ` when a bucket is removed, all policies `

**Comment by vadmeste** [comment]
@poornas after introducing rb command, do you expect that rm is still able to remove objects from different buckets at the same time ?

**Comment by poornas** [comment]
> @poornas after introducing rb command, do you expect that rm is still able to remove objects from different buckets at the same time ?

@vadmeste, yes - both the commands below will remove objects from different buckets, except that mc rb will delete bucket in addition. 
```
➜  mc git:(rb) mc rb --force s3/tbucket11 s3/tbucket13
Bucket removed successfully `s3/tbucket11`.
Bucket removed successfully `s3/tbucket13`.
```
and 
```
➜  mc git:(rb) mc rm --r --force s3/tbucket11 s3/tbucket13
```

**Comment by vadmeste** [comment]
@poornas I see, but I also mean this:
     `mc rm alias/`

does it also need to traverse several buckets and remove all objects in them ?

**Comment by poornas** [comment]
> @poornas I see, but I also mean this:
> `mc rm alias/`
> 
> does it also need to traverse several buckets and remove all objects in them ?
@vadmeste , that functionality also remains the same as before.

File: cmd/rb-main.go [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+		if !isForce {
+			if isNamespaceRemoval(targetURL) {
+				fatalIf(errDummy().Trace(),
+					"This operation results in **site-wide** removal of buckets. If you are really sure, retry this command with ‘--force’ flags.")
+			}
+			if !isEmpty {
+				fatalIf(errDummy().Trace(), "`"+targetURL+"` is not empty. Retry this command with ‘--force’ flag if you want to remove `"+targetURL+"` and all its contents")
+			}
+		}
+		if e := deleteBucket(targetURL); e == nil {
```
Comment (vadmeste) [inline cmd/rb-main.go]: Why we don't print errors (`err != nil` case) ?

File: cmd/rb-main.go [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+	for _, targetURL := range ctx.Args() {
+		// Instantiate client for URL.
+		clnt, err := newClient(targetURL)
+		if err != nil {
+			errorIf(err.Trace(targetURL), "Invalid target `"+targetURL+"`.")
+			cErr = exitStatus(globalErrorExitStatus)
+			continue
+		}
+		isEmpty := true
+		for range clnt.List(true, false, DirNone) {
```
Comment (vadmeste) [inline cmd/rb-main.go]: I think you need to pass DirFirst here, we can have some empty directories sometimes

Comment (poornas) [reply cmd/rb-main.go]: @vadmeste, it is my understanding that DirFirst option would allow listing directory entries ahead of files. Here, the purpose is to ascertain if the bucket in targeturl is empty , hence listing with DirNone. fwiw, I tested with empty directory and it behaves correctly.
```
➜  minio git:(master) mc ls myminio/newb --r
[2019-02-13 23:01:58 PST]     0B em3/
➜  minio git:(master) mc mb myminio/newb    
Bucket created successfully `myminio/newb`.
➜  minio git:(master) mc ls myminio/newb --r    
[2019-02-13 23:03:47 PST]     0B em3/
➜  minio git:(master) mc rb myminio/newb
mc: <ERROR> `myminio/newb` is not empty. Retry this command with ‘--force’ flag if you want to remove `myminio/newb` and all its contents 
➜  minio git:(master) mc rb myminio/newb --force
Bucket removed successfully `myminio/newb`.
```

File: cmd/rb-main.go (hunk 3 above)
Comment (poornas) [reply cmd/rb-main.go]: @vadmeste,changed this to print error also.

File: cmd/client-s3.go [hunk 5]
```diff
@@ -36,7 +36,7 @@ import (
 
 	"github.com/minio/mc/pkg/httptracer"
 	"github.com/minio/mc/pkg/probe"
-	"github.com/minio/minio-go"
+	minio "github.com/minio/minio-go"
```
Comment (kannappanr) [inline cmd/client-s3.go]: @poornas This code has not been used anywhere else in this file. Is this change required?

File: docs/minio-client-complete-guide.md [hunk 6]
```diff
@@ -255,7 +256,7 @@ Skip SSL certificate verification.
 | [**diff** - Diff buckets](#diff) |[**policy** - Set public policy on bucket or prefix](#policy)  |[**session** - Manage saved sessions](#session) |
 | [**config** - Manage config file](#config)  | [**watch** - Watch for events](#watch)  | [**event** - Manage events on your buckets](#event)  |
 | [**update** - Manage software updates](#update)  | [**version** - Show version](#version)  | [**stat** - Stat contents of objects and folders](#stat) |
-| [**head** - Display first 'n' lines of an object](#head) | [**sql** - Run sql queries on objects](#sql) | |
+| [**head** - Display first 'n' lines of an object](#head) | [**sql** - Run sql queries on objects](#sql) | [**rb** - Remove a bucket](#rb) |
```
Comment (kannappanr) [inline docs/minio-client-complete-guide.md]: It is better to place `rb` next to `mb`

File: cmd/client-s3.go (hunk 5 above)
Comment (harshavardhana) [reply cmd/client-s3.go]: This is perhaps automatic with `goimports` through the editor.

File: docs/minio-client-complete-guide.md (hunk 6 above)
Comment (poornas) [reply docs/minio-client-complete-guide.md]: done @kannappanr
